iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp

```diff
@@ -0,0 +1,1149 @@
+    assert(it != commandBuffers.end() &&
+           "command buffer must have been registered during conversion");
+    return it->second;
+  }
+
+ private:
+  // Ops within stream.cmd.execute ops -> !hal.command_buffer.
+  DenseMap<Operation *, Value> commandBuffers;
+};
+
+template <typename OpT>
+struct StreamConversionPattern : public OpConversionPattern<OpT> {
+  StreamConversionPattern(std::shared_ptr<StreamConversionMapping> mapping,
+                          TypeConverter &typeConverter, MLIRContext *context,
+                          PatternBenefit benefit = 1)
+      : OpConversionPattern<OpT>(typeConverter, context, benefit),
+        mapping(std::move(mapping)) {}
+
+  std::shared_ptr<StreamConversionMapping> mapping;
+};
+
+struct ResourceAllocOpPattern
+    : public StreamConversionPattern<IREE::Stream::ResourceAllocOp> {
+  using StreamConversionPattern::StreamConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Stream::ResourceAllocOp allocOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto allocator = lookupAllocatorFor(allocOp, rewriter);
+    auto bufferType = rewriter.getType<IREE::HAL::BufferType>();
+
+    SmallVector<Value> results;
+    for (auto it : llvm::zip(allocOp.results(), allocOp.storage_sizes())) {
+      auto resourceResult = std::get<0>(it);
+      auto resourceType =
+          resourceResult.getType().cast<IREE::Stream::ResourceType>();
+      auto storageSize = std::get<1>(it);
+
+      auto memoryTypes = IREE::HAL::MemoryTypeBitfield::None;
+      auto bufferUsage = IREE::HAL::BufferUsageBitfield::None;
+      switch (resourceType.getLifetime()) {
+        default:
+          return allocOp.emitOpError()
+                 << "unsupported resource lifetime: "
+                 << IREE::Stream::stringifyLifetime(resourceType.getLifetime());
+        case IREE::Stream::Lifetime::Constant:
+          // Device local; copies required to get into external resources.
+          memoryTypes =
+              memoryTypes | IREE::HAL::MemoryTypeBitfield::DeviceLocal;
+          bufferUsage = bufferUsage | IREE::HAL::BufferUsageBitfield::Constant;
+          break;
+        case IREE::Stream::Lifetime::Variable:
+          // Device local; copies required to get into external resources.
+          memoryTypes =
+              memoryTypes | IREE::HAL::MemoryTypeBitfield::DeviceLocal;
+          break;
+        case IREE::Stream::Lifetime::External:
+          // #yolo; these come from/go to outside the program.
+          // Today we assume they are device-local|host-visible just for
+          // practical purposes but that does not have to be true. We really
+          // want this to be something we analyze and handle on the edges
+          // (transfering devices/etc if needed).
+          memoryTypes = memoryTypes |
+                        IREE::HAL::MemoryTypeBitfield::DeviceLocal |
+                        IREE::HAL::MemoryTypeBitfield::HostVisible;
```

Comment:
If the memory is not `device-local|host-visible` what will happen? Anything we can do now to give a better error message until this is generalized?

Reply:
the hal.buffer.assert op/export added after this give a nice message here on input if the memory type or usage bits don't match 🎉